From 989b5e5169bd07a1d82383612280f696ff5933fc Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Sat, 2 Mar 2024 21:19:25 +0100
Subject: [PATCH 14/32] [HACK] mfmediaengine: Do not send
 MF_MEDIA_ENGINE_EVENT_ERROR.

Breaks AVPro Video (used by VRChat) when a IMFMediaEngine::GetError call inside the event handler doesn't return an
IMFMediaError instance because the source is already shut down. The event handler code tries to call a function on the
non-existant IMFMediaError instance, which segfaults.

This should be fixed properly. In the meantime, this hack works around the problem.
---
 dlls/mfmediaengine/main.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/dlls/mfmediaengine/main.c b/dlls/mfmediaengine/main.c
index 3e05626f238..e462a918f96 100644
--- a/dlls/mfmediaengine/main.c
+++ b/dlls/mfmediaengine/main.c
@@ -1397,11 +1397,21 @@ static HRESULT WINAPI media_engine_load_handler_Invoke(IMFAsyncCallback *iface,
     }
     else
     {
-        engine->network_state = MF_MEDIA_ENGINE_NETWORK_NO_SOURCE;
-        engine->error_code = MF_MEDIA_ENGINE_ERR_SRC_NOT_SUPPORTED;
-        engine->extended_code = hr;
-        IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_ERROR, engine->error_code,
-                engine->extended_code);
+        static unsigned int once;
+        char sgi[64];
+
+        if (GetEnvironmentVariableA("SteamGameId", sgi, sizeof(sgi)) && !strcmp(sgi, "438100"))
+        {
+            if (!once++) FIXME("HACK: not sending MF_MEDIA_ENGINE_EVENT_ERROR event.\n");
+        }
+        else
+        {
+            engine->network_state = MF_MEDIA_ENGINE_NETWORK_NO_SOURCE;
+            engine->error_code = MF_MEDIA_ENGINE_ERR_SRC_NOT_SUPPORTED;
+            engine->extended_code = hr;
+            IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_ERROR, engine->error_code,
+                    engine->extended_code);
+        }
     }
 
     LeaveCriticalSection(&engine->cs);
-- 
2.44.0

